### ISSUE TITLE
SendAsync  SemaphoreSlim Exception

### ISSUE THREAD

**Issue by ygr757326** [kerryjiang/SuperSocket#748]
![1725873682653](https://github.com/user-attachments/assets/4d4512af-cd7c-43ba-811c-f46fbda821aa)

MyPackageHandler SendAsync error：Adding the specified count to the semaphore would cause it to exceed its maximum count.
System.Threading.SemaphoreFullException: Adding the specified count to the semaphore would cause it to exceed its maximum count.
   at System.Threading.SemaphoreSlim.Release(Int32 releaseCount)
   at SuperSocket.Connection.PipeConnectionBase.SendAsync[TPackage](IPackageEncoder`1 packageEncoder, TPackage package, CancellationToken cancellationToken)

**Comment by wj8400684** [comment]
You cannot use async within a foreach loop.

**Comment by kerryjiang** [comment]
That's strange, I will take a look at this problem. That would be great if you have a unit test to reproduce this problem.

**Comment by Helder-Henderson** [comment]
Good morning

I had same problem yesterday... 

In my cenario i throw some exceptions on Command - ExecuteAsync. 

2024-09-21 23:24:53.174 -03:00 [ERR] Session[33bb9838-4f36-4dc5-9461-3607db5a2c90]: session exception.
SuperSocket.PackageHandlingException`1[SuperSocket.WebSocket.WebSocketPackage]: Session 33bb9838-4f36-4dc5-9461-3607db5a2c90 got an error when handle a package.
 ---> System.InvalidOperationException: Writing is not allowed after writer was completed.
   at System.IO.Pipelines.ThrowHelper.ThrowInvalidOperationException_NoWritingAllowed()
   at SuperSocket.WebSocket.WebSocketEncoder.WriteHead(IBufferWriter`1 writer, Int64 length, Int32& headLen)
   at SuperSocket.WebSocket.WebSocketEncoder.EncodeFinalFragment(IBufferWriter`1 writer, Byte opCode, ReadOnlySpan`1 text, Encoder encoder, ArraySegment`1 unwrittenBytes)
   at SuperSocket.WebSocket.WebSocketEncoder.Encode(IBufferWriter`1 writer, WebSocketPackage pack)
   at SuperSocket.Connection.PipeConnectionBase.WritePackageWithEncoder[TPackage](IBufferWriter`1 writer, IPackageEncoder`1 packageEncoder, TPackage package)
   at SuperSocket.Connection.PipeConnectionBase.SendAsync[TPackage](IPackageEncoder`1 packageEncoder, TPackage package, CancellationToken cancellationToken)
   at GerenciadorFacial.Commands.ReceberLog.ExecuteAsync(EquipamentoSession session, CustomDataPackage package, CancellationToken cancellationToken) in C:\dev\MDAcesso\ProjetoAugusto\GerenciadorFacial\Commands\ReceberLog.cs:line 32
   at SuperSocket.Command.AsyncCommandWrap`4.ExecuteAsync(TAppSession session, TPackageInfo package, CancellationToken cancellationToken)
   at SuperSocket.Command.CommandMiddleware`3.CommandSet`1.ExecuteAsync(IAppSession session, TPackageInfo package, CancellationToken cancellationToken)
   at SuperSocket.Command.CommandMiddleware`3.HandlePackage(IAppSession session, TPackageInfo package, CancellationToken cancellationToken)
   at SuperSocket.WebSocket.Server.WebSocketPackageHandler.Handle(IAppSession session, WebSocketPackage package, CancellationToken cancellationToken)
   at SuperSocket.Server.PackageHandlingSchedulerBase`1.HandlePackageInternal(IAppSession session, TPackageInfo package, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---

**Comment by Helder-Henderson** [comment]
@kerryjiang is possible to implement one sample using JsonAsyncCommand ?

Basic basic, only for understand it.

**Comment by kerryjiang** [comment]
Found the problem and working on a fix.

**Comment by kerryjiang** [comment]
Pushed a fix.

With this fix, you would get TaskCancellationException in the same situation.

I suggest you check the cancellationToken is triggered before you call SendAsync(*).

```
if (!cancellationToken.IsCancellationRequested)
    await session.SendAsync(xxxx, cancellationToken);

```

cancellationToken.IsCancellationRequested == true means the connection is closed or is being closed.

**Comment by kerryjiang** [comment]
> @kerryjiang is possible to implement one sample using JsonAsyncCommand ?
> 
> Basic basic, only for understand it.

@Helder-Henderson Do you use token in async json command? Token allows your client to talk with server in this order:

Request A, Request B, Request C, Response C, Response A, Response B.

**Comment by Helder-Henderson** [comment]
No no, sorry, I found implementation about JsonCommands on unit tests... 

My questions was not related to the issue, sorry !

**Comment by kerryjiang** [comment]
Released SuperSocket 2.0.0-beta.27 to fix this issue:
https://www.nuget.org/packages/SuperSocket/2.0.0-beta.27
